pkg/syncer/admissionhandler/admissionhandler.go

```diff
@@ -0,0 +1,142 @@
+package admissionhandler
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	admissionv1 "k8s.io/api/admission/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+
+	"sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/logger"
+)
+
+type (
+	parameterSet map[string]struct{}
+)
+
+// Has checks if specified paramName present in the paramSet
+func (paramSet parameterSet) Has(paramName string) bool {
+	_, ok := paramSet[paramName]
+	return ok
+}
+
+var (
+	webhookServerPort = 8443
+	server            *http.Server
+)
+
+// StartWebhookServer starts the webhook server
+func StartWebhookServer(ctx context.Context, cert string, key string, port string) error {
+	log := logger.GetLogger(ctx)
+	certs, err := tls.X509KeyPair([]byte(cert), []byte(key))
+	if err != nil {
+		log.Errorf("failed to load key pair: %v", err)
+		return err
+	}
+	if port == "" {
+		port = fmt.Sprintf("%v", webhookServerPort)
+	}
+	server = &http.Server{
+		Addr:      fmt.Sprintf(":%v", port),
+		TLSConfig: &tls.Config{Certificates: []tls.Certificate{certs}},
+	}
+	// define http server and server handler
+	mux := http.NewServeMux()
+	mux.HandleFunc("/validate-storageclass", validationHandler)
+	server.Handler = mux
+
+	// start webhook server
+	log.Debugf("starting webhook server on port: %v", port)
+	if err = server.ListenAndServeTLS("", ""); err != nil {
+		log.Errorf("failed to listen and serve webhook server: %v", err)
+		return err
+	}
+	return nil
+}
+
+// validationHandler is the handler for webhook http multiplexer to help validate resources
+// depending on the URL validation of AdmissionReview will be redirected to appropriate validation function
+func validationHandler(w http.ResponseWriter, r *http.Request) {
+	var body []byte
+	ctx, log := logger.GetNewContextWithLogger()
+	if r.Body != nil {
+		if data, err := ioutil.ReadAll(r.Body); err == nil {
+			body = data
+		}
+	}
+	if len(body) == 0 {
+		log.Error("received empty request body")
+		http.Error(w, "received empty request body", http.StatusBadRequest)
+		return
+	}
+	log.Debugf("Received request")
+	// verify the content type is accurate
+	contentType := r.Header.Get("Content-Type")
+	if contentType != "application/json" {
+		log.Errorf("Content-Type=%s, expect application/json", contentType)
+		http.Error(w, "invalid Content-Type, expect `application/json`", http.StatusUnsupportedMediaType)
+		return
+	}
+
+	var admissionResponse *admissionv1.AdmissionResponse
+	ar := admissionv1.AdmissionReview{}
+	codecs := serializer.NewCodecFactory(runtime.NewScheme())
+	deserializer := codecs.UniversalDeserializer()
+	if _, _, err := deserializer.Decode(body, nil, &ar); err != nil {
+		log.Errorf("Can't decode body: %v", err)
+		admissionResponse = &admissionv1.AdmissionResponse{
+			Result: &metav1.Status{
+				Message: err.Error(),
+			},
+		}
+	} else {
+		if r.URL.Path == "/validate-storageclass" {
```

Comment:
lets keep this generic path. Once you decode, determine if it is a storage class object and call validateStorageClass.

Reply:
done.

```
              if r.URL.Path == "/validate" {
			log.Debugf("request URL path is /validate")
			log.Debugf("admissionReview: %+v", ar)
			switch ar.Request.Kind.Kind {
			case "StorageClass":
				admissionResponse = validateStorageClass(ctx, &ar)
			default:
				admissionResponse = &admissionv1.AdmissionResponse{
					Allowed: true,
				}
			}
			log.Debugf("admissionResponse: %+v", admissionResponse)
		}
```